quic-forward-proxy/src/outbound/tpu_connection_manager.rs

```diff
@@ -0,0 +1,283 @@
+            exit_signal.clone(),
+            max_number_of_connections,
+            max_uni_stream_connections,
+        );
+
+        tokio::spawn(
+            Self::listen(
+                connection_pool.clone(),
+                broadcast_receiver,
+                exit_signal.clone(),
+                self.identity,
+                addr.clone(),
+                self.last_used.clone(),
+            ));
+
+    }
+}
+
+pub struct TpuConnectionManager {
+    endpoints: RotatingQueue<Endpoint>,
+    identity_to_active_connection: Arc<DashMap<Pubkey, Arc<ActiveConnection>>>,
+    // channel to communicate to the active connections threads
+    broadcast_sender: Arc<Sender<BroadcastMessage>>,
+}
+
+impl TpuConnectionManager {
+    pub fn send_transaction(&self, transaction: Vec<u8>) {
+        self.broadcast_sender.send(BroadcastMessage::Transaction(transaction))
+            .expect("failed to send to broadcast");
+    }
+}
+
+impl TpuConnectionManager {
+    pub async fn new(
+        certificate: rustls::Certificate,
+        key: rustls::PrivateKey,
+        fanout: usize,
+    ) -> Self {
+
+        let broadcast_sender = {
+            let (sender, _) = tokio::sync::broadcast::channel::<BroadcastMessage>(MAXIMUM_TRANSACTIONS_IN_QUEUE);
+            Arc::new(sender.clone())
+        };
+
+        let identity_to_active_connection = Arc::new(DashMap::new());
+
+        tokio::spawn(
+            Self::cleanup_unused_connections(
+            identity_to_active_connection.clone(),
+            broadcast_sender.clone(),
+            ));
+
+        let number_of_clients = fanout * 2;
+        Self {
+            endpoints: RotatingQueue::new(number_of_clients, || {
+                QuicConnectionUtils::create_endpoint(certificate.clone(), key.clone())
+            }),
+            identity_to_active_connection,
+            broadcast_sender,
+        }
+    }
+
+    #[tracing::instrument(skip_all, level = "warn")]
+    pub async fn update_connections(
+        &self,
+        requested_connections: &HashMap<Pubkey, SocketAddr>,
+        max_uni_stream_connections: usize,
+        connection_parameters: QuicConnectionParameters,
+    ) {
+        NB_CONNECTIONS_TO_KEEP.set(requested_connections.len() as i64);
+        for (tpu_identity, tpu_addr) in requested_connections {
+            if self.identity_to_active_connection.get(tpu_identity).is_some() {
+                continue;
+            }
+            debug!("add active connection for {}, {}", tpu_identity, tpu_addr);
+            let active_connection = ActiveConnection::new(
+                self.endpoints.clone(),
+                *tpu_addr,
+                *tpu_identity,
+                connection_parameters,
+            );
+
+            let broadcast_receiver = self.broadcast_sender.subscribe();
+
+            active_connection.start_listening(broadcast_receiver, max_uni_stream_connections);
+            self.identity_to_active_connection.insert(
```

Comment:
Connections are never removed from this map. They are always added.

Reply:
eh! don't scare me ;) - look, I guess you are wrong:

> 2023-09-29T15:08:05.018721Z  INFO update_connections: solana_lite_rpc_quic_forward_proxy::outbound::tpu_connection_manager: identity_to_active_connection: 9 
...
2023-09-29T15:08:13.737999Z  INFO update_connections: solana_lite_rpc_quic_forward_proxy::outbound::tpu_connection_manager: identity_to_active_connection: 3